R/derive_params_growth_age.R

```diff
@@ -0,0 +1,235 @@
+#'     AGECURU = ifelse(AGECUR_D >= 365.25 * 2, CURU_M, CURU_D)
+#'   )
+#'
+#' #' metadata is in months
+#' cdc_meta_criteria <- admiralpeds::cdc_wtage %>%
+#'   mutate(
+#'     age_unit = "months",
+#'     SEX = ifelse(SEX == 1, "M", "F")
+#'   )
+#'
+#' #' metadata is in days
+#' who_meta_criteria <- bind_rows(
+#'   (admiralpeds::who_wt_for_age_boys %>%
+#'     mutate(
+#'       SEX = "M",
+#'       age_unit = "days"
+#'     )
+#'   ),
+#'   (admiralpeds::who_wt_for_age_girls %>%
+#'     mutate(
+#'       SEX = "F",
+#'       age_unit = "days"
+#'     )
+#'   )
+#' ) %>%
+#'   rename(AGE = Day)
+#'
+#' criteria <- bind_rows(
+#'   cdc_meta_criteria,
+#'   who_meta_criteria
+#' ) %>%
+#'   rename(AGEU = age_unit)
+#'
+#' derive_params_growth_age(
+#'   advs,
+#'   sex = SEX,
+#'   age = AGECUR,
+#'   age_unit = AGECURU,
+#'   meta_criteria = criteria,
+#'   parameter = VSTESTCD == "WEIGHT",
+#'   set_values_to_sds = exprs(
+#'     PARAMCD = "WTASDS",
+#'     PARAM = "Weight-for-age z-score"
+#'   ),
+#'   set_values_to_pctl = exprs(
+#'     PARAMCD = "WTAPCTL",
+#'     PARAM = "Weight-for-age percentile"
+#'   )
+#' )
+derive_params_growth_age <- function(dataset,
+                                     sex,
+                                     age,
+                                     age_unit,
+                                     meta_criteria,
+                                     parameter,
+                                     set_values_to_sds = NULL,
+                                     set_values_to_pctl = NULL) {
+  # Apply assertions to each argument to ensure each object is appropriate class
+  # need to add assertion for dataset that checks columns in sex, age, age_unit are present
+  assert_data_frame(meta_criteria, required_vars = exprs(SEX, AGE, AGEU, L, M, S))
+  assert_expr(enexpr(parameter))
+  assert_varval_list(set_values_to_sds, optional = TRUE)
+  assert_varval_list(set_values_to_pctl, optional = TRUE)
+
+  # create a unified join naming convention, hard to figure out in by argument
+  dataset <- dataset %>%
+    mutate(
+      sex_join := {{ sex }},
+      ageu_join := {{ age_unit }}
+    )
+
+  # Process metadata
+  # Metadata should contain SEX, AGE, AGEU, L, M, S
+  # Processing the data to be compatible with the dataset object
+  processed_md <- meta_criteria %>%
+    arrange(SEX, AGEU, AGE) %>%
+    group_by(SEX, AGEU) %>%
+    mutate(next_age = lead(AGE)) %>% # needed for the join and filter later creates [x, y) range
+    rename(
+      sex_join = SEX,
+      prev_age = AGE,
+      ageu_join = AGEU
+    )
+
+  # Merge the dataset that contains the vs records and filter the L/M/S that fit the appropriate age
+  # To parse out the appropriate age, create [x, y) brackets using a prev_age <= AGECUR < next_age
+  added_records <- dataset %>%
+    filter(!!enexpr(parameter)) %>%
+    left_join(.,
+      processed_md,
+      by = c("sex_join", "ageu_join"),
```

Comment:
I decided by joining the entire metadataset by sex and age unit then filtering the correctly mapped L/M/S was more appropriate than creating some sort of weird fuzzy join or join_by statement using within() or overlaps(), reason this came up was imagine a dataset that has a

24 days old and 24.6 months old patient in the same study

we run into a couple issues:
- desired age units is not homogenous in the study
- metadata begins to have "gaps" between ages
- the 24 days old patient should have a pretty easy/direct mapping to a metadata entry appropriate for 24 days old, at least in the WHO data its per day, which is the smallest denomination anyway so thats fine but
- the 24.6 months old patient is in the range between 24.5 and 25.5 months old for the CDC data

I played around having the function only be able to map one age unit at a time, but running the function twice felt clunky, the way I've currently programmed it left a decent amount of flexibility depending how the user wants to configure their metadata